modules/example/src/smithy4s/example/typeclass/EqInterpreter.scala

```diff
@@ -0,0 +1,212 @@
+package smithy4s.example.typeclass
+
+import cats.implicits.{catsSyntaxEq, toContravariantOps}
+import smithy4s.{Schema, _}
+import cats.kernel.Eq
+import smithy4s.capability.EncoderK
+import smithy4s.schema._
+
+import java.util.UUID
+
+object EqInterpreter extends CachedSchemaCompiler.Impl[Eq] {
+
+  protected type Aux[A] = Eq[A]
+
+  def fromSchema[A](
+      schema: Schema[A],
+      cache: Cache
+  ): Eq[A] = {
+      schema.compile(EqSchemaVisitor)
+  }
+
+}
+
+trait EqInstances {
+  implicit def arrayEq[A: Eq]: Eq[Array[A]] = (x: Array[A], y: Array[A]) =>
+    x.zip(y).forall { case (a, b) => a === b }
+
+  implicit def indexedSeq[A: Eq]: Eq[IndexedSeq[A]] =
+    Eq[Seq[A]].contramap(_.toSeq)
+
+  implicit val byteArrayEq: Eq[ByteArray] = (x: ByteArray, y: ByteArray) =>
+    Eq[Array[Byte]].contramap[ByteArray](_.array).eqv(x, y)
+  implicit val documentEq: Eq[Document] =
+    implicitly[Eq[String]].contramap(_.show)
+  implicit val timeStampEq: Eq[Timestamp] =
+    implicitly[Eq[Long]].contramap(_.epochSecond)
+
+}
+object EqInstances extends EqInstances
+
+object EqSchemaVisitor extends SchemaVisitor[Eq] { self =>
+  import EqInstances._
+
+  override def primitive[P](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: Primitive[P]
+  ): Eq[P] = primitiveEq(tag)
+
+  override def collection[C[_], A](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: CollectionTag[C],
+      member: Schema[A]
+  ): Eq[C[A]] = {
+    implicit val memberEq: Eq[A] = self(member)
+    tag match {
+      case CollectionTag.ListTag       => Eq[List[A]]
+      case CollectionTag.SetTag        => Eq[Set[A]]
+      case CollectionTag.VectorTag     => Eq[Vector[A]]
+      case CollectionTag.IndexedSeqTag => Eq[IndexedSeq[A]]
+    }
```

Comment:
thought: we could have a `CollectionTag.deriving` in a similar spirit to `Primitive.deriving`